src/filesystem/impls/filer/lib/ImageResizer.js

```diff
@@ -0,0 +1,88 @@
+/*jslint vars: true, plusplus: true, devel: true, nomen: true, regexp: true, indent: 4, maxerr: 50 */
+/*global define */
+define(function (require, exports, module) {
+    "use strict";
+
+    var pica = require("Pica")();
+
+    var Content = require("filesystem/impls/filer/lib/content");
+    var Path = require("filesystem/impls/filer/FilerUtils").Path;
+    var decodePath = require("filesystem/impls/filer/FilerUtils").decodePath;
+    var base64ToBuffer = require("filesystem/impls/filer/FilerUtils").base64ToBuffer
+
+    function ImageResizer(path, data) {
+        this.path = decodePath(path);
+        this.type = Content.mimeFromExt(Path.extname(this.path));
+        this.blob = new Blob([data], {type: this.type});
+        this.url = URL.createObjectURL(this.blob);
+        this.canvas = document.createElement('canvas');
+    }
+    ImageResizer.prototype.cleanup = function() {
+        URL.revokeObjectURL(this.url);
+    };
+    ImageResizer.prototype.resize = function(callback) {
+        var self = this;
+        var scalingFactor = 0.75;
+        var buffer;
+        var passes = 0;
+
+        function finish(err, buffer) {
+            self.cleanup();
+            callback(err, buffer);
+        }
+
+        function resizePass() {
+            console.log("resizePass", passes, scalingFactor);
+
+            if(passes++ > 5) {
+                finish(null, buffer);
+            }
+
+            self.canvas.width = img.width * scalingFactor;
+            self.canvas.height = img.height * scalingFactor;
+
+            console.log("width", self.canvas.width, "height", self.canvas.height);
+
+            pica.resize(img, self.canvas)
+            .then(function() {
+                var base64Str = self.canvas.toDataURL(self.type).split(',')[1];
+                buffer = base64ToBuffer(base64Str);
+
+                console.log("result", buffer.length);
+
+                if(isImageTooLarge(buffer.length)) {
+                    scalingFactor /= 1.25;
```

Comment:
For the iteration I would recommend a binary search:
```
scale = 1 // initial scale factor of 1 to deal with "only huge because huge dpi" images
step = scaling / 2
target = 250kb
errorThreshold = 20kb
passes = 0
maxPasses = 5

do {
  passes++

  result = resize(image, scale, step)
  actual = result.size

  // too big?
  if (actual > target + errorThreshold) {
    scale = scale - step
    step /= 2
    repeat run
  }

  // smaller than necessary?
  else if (actual < target - errorThreshold) {
    scale = scale + step
    step /= 2
    repeat run
  }
  return result
} while (passes < maxPasses)
```
This lets us converge "up" as well as down, so that we don't give someone a 125000 byte file because 250001 was too big.